src/library/scala/collection/concurrent/TrieMap.scala

```diff
@@ -943,12 +943,17 @@ extends scala.collection.concurrent.Map[K, V]
 
   override def stringPrefix = "TrieMap"
 
+  // This is needed to instantiate the updater in this class, because it is accessing
+  // a private field, and then leak it to the companion object for storage
+  private def updater(): AtomicReferenceFieldUpdater[TrieMap[_, _], AnyRef] =
+    AtomicReferenceFieldUpdater.newUpdater(classOf[TrieMap[_, _]], classOf[AnyRef], "root")
 }
 
-
 object TrieMap extends MutableMapFactory[TrieMap] {
   val inodeupdater = AtomicReferenceFieldUpdater.newUpdater(classOf[INodeBase[_, _]], classOf[MainNode[_, _]], "mainnode")
 
+  private val rootupdater = new TrieMap[Any, Any](null, null, null, null).updater()
```

Comment:
The binary compatibility is real:

```
$ javap -cp /Users/szeiger/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.12.0.jar scala.collection.concurrent.TrieMap | grep "TrieMap("
  public scala.collection.concurrent.TrieMap(scala.util.hashing.Hashing<K>, scala.math.Equiv<K>);
  public scala.collection.concurrent.TrieMap();

$ javap -cp /Users/szeiger/code/scala/build/pack/lib/scala-library.jar scala.collection.concurrent.TrieMap | grep "TrieMap("
  public scala.collection.concurrent.TrieMap(java.lang.Object, java.util.concurrent.atomic.AtomicReferenceFieldUpdater<scala.collection.concurrent.TrieMap<?, ?>, java.lang.Object>, scala.util.hashing.Hashing<K>, scala.math.Equiv<K>);
  public scala.collection.concurrent.TrieMap(scala.util.hashing.Hashing<K>, scala.math.Equiv<K>);
  public scala.collection.concurrent.TrieMap();
```

This here is the culprit. This constructor used to be JVM-private but calling it from the companion object makes it JVM-public.